add clear.js and length.js to stored procedures samples

dwhieb opened the pull request.

dwhieb commented:
Just following up - any movement on this PR? Thanks!

rnagpal commented:
@dwhieb : Sorry for the delay. I've asked the respective owners to take a look at it.

dwhieb commented:
No worries - thanks!

emilylawton commented:
Thanks, @dwhieb! These samples are great. Our bank of stored procedure samples currently includes Count.js which performs a document count with an optional filter and bulkDelete.js which deletes documents for a given query. What additional functionality did you want to add with clear.js and length.js?

dwhieb commented:
Well for both sprocs I added the option to pass a key and value to filter on. And for count.js I used the collection's .filter() method, which seemed a little more straightforward since I was in fact filtering.

I'd be happy to incorporate these changes into the sprocs already in the repo if that'd be better. And in retrospect I think passing a filter object rather than two arguments would be a better way to structure the filter option.

koltachev commented:
Thanks for reaching out!
I think since these are using underscore API it would make sense to keep them this way.
There is a couple of issues which is not a big deal -- I'll add a few comments.

koltachev commented on samples/stored-procedures/clear.js:
```diff
@@ -0,0 +1,121 @@
+/**
```
Can this be renamed to bulkDelete_underscoreAPI.js to make it clear this is similar to bulkDelete.js but using underscore API?

dwhieb replied:
780ca9c1d0f07b4eb50c78a6030d3d74441b0296

emilylawton commented on samples/stored-procedures/clear.js:
```diff
@@ -0,0 +1,121 @@
+ * @param  {String} filterValue = 'session'   The value that a document's filter key must have to be deleted
+ * @return {responseBody}
+ */
+function clear(filterOn, filterValue) {
```
,mkhkhkj

koltachev commented on samples/stored-procedures/clear.js:
```diff
@@ -0,0 +1,121 @@
+    };
+
+    // filter the collection for sessions using the filter function
+    const accepted = __.filter(filter, { continuation: continuationToken }, handler);
```
Perf: the predicate function needs to be inlined in order to take advantage of index (otherwise it will be full scan).

dwhieb replied:
f0f61d2ea5f3fa03f6925e11fcf7d9f1b2d44007

koltachev commented on samples/stored-procedures/length.js:
```diff
@@ -0,0 +1,90 @@
+/**
+ * A stored procedure for Azure DocumentDB which gets a count of the session documents using the .filter() method of the collection.
+ * @function
+ * @param  {String} filterOn = 'type'         The key to filter documents on for deletion.
```
Should say for counting instead of deletion.

dwhieb replied:
afbbec747e479814d5e9e9537258a88870be1193

koltachev commented on samples/stored-procedures/length.js:
```diff
@@ -0,0 +1,90 @@
+/**
```
Can this be renamed to bulkDelete_underscoreAPI.js to make it clear this is similar to bulkDelete.js but using underscore API?

koltachev replied:
Can this be renamed to count_underscoreAPI.js to make it clear this is similar to bulkDelete.js but using underscore API?

koltachev replied:
Can this be renamed to count_underscoreAPI.js to make it clear this is similar to count.js but using underscore API?

dwhieb replied:
be725a2ca369d984e0f5654a55ad2b3d07419ebd

koltachev commented on samples/stored-procedures/length.js:
```diff
@@ -0,0 +1,90 @@
+    };
+
+    // filter the collection for sessions using the filter function
+    const accepted = __.filter(filter, { continuation: continuationToken }, handler);
```
Perf: the predicate function needs to be inlined in order to take advantage of index (otherwise it will be full scan).

dwhieb replied:
b0f33048fb2aa988a6fcbe37141dd83dd6654e4f

koltachev commented on samples/stored-procedures/length.js:
```diff
@@ -0,0 +1,90 @@
+
+}
+
+module.exports = length;
```
I don't think module is currently supported (will be soon).

dwhieb replied:
1763c9925953eec28a73f51b625a94c060327f9f

koltachev commented on samples/stored-procedures/length.js:
```diff
@@ -0,0 +1,90 @@
+     * @param  {String} info.continuation   The continuation token, if any was passed
+     * @return {responseBody}
+     */
+    const handler = function handler(err, sessions, info) {
```
Can we rename 'sessions' parameter to something more generic, like docs, or documents, or docFeed?

dwhieb replied:
3c157495761211e3d9738182fb6047c5907fe8c9

koltachev commented on samples/stored-procedures/clear.js:
```diff
@@ -0,0 +1,121 @@
+  /**
+   * Recursively deletes each session in an array, and then attempts to get more to delete
+   * @function
+   * @param  {Array} sessions  The array of session documents to delete
```
can we rename sessions for function name and function parameter to more generic? Like deleteDocuments etc.

dwhieb replied:
3c157495761211e3d9738182fb6047c5907fe8c9

dwhieb commented:
@koltachev Ok I've added my revisions to the PR and addressed each of your comments (I noted the commit for each), as well as changed the way the filter parameter is structured, so that the sprocs now take a single (optional) filter object as an argument.

mkolt commented on samples/stored-procedures/bulkDelete_underscoreAPI.js:
```diff
@@ -0,0 +1,114 @@
+    // NB: The filter function must be inlined in order to take advantage of index
+    // (otherwise it will be a full scan).
+    const accepted = __.filter(function filter(doc) {
+      return Object.keys(filterObj).every(function checkProperty(filterKey) {
```
Function calls are not yet supported for optimized index, so this will work but again, will be full scan - slow. This scenario will work when we have __.where implemented. For now, I guess, let's change the sample to the version it was before -- simple filtering by property name and property value passed from outside.

mkolt replied:
Function calls are not yet supported for optimized index, so this will work but again, will be full scan - slow. This scenario will work when we have __.where implemented. For now, I guess, let's change the sample to the version it was before -- simple filtering by property name and property value passed from outside. Same for the count sproc.

dwhieb replied:
https://github.com/Azure/azure-documentdb-js-server/pull/5/commits/a449eb006260af52594a3b8aad6d89ab4e6a08fb

dwhieb replied:
https://github.com/Azure/azure-documentdb-js-server/pull/5/commits/a449eb006260af52594a3b8aad6d89ab4e6a08fb

mkolt commented on samples/stored-procedures/bulkDelete_underscoreAPI.js:
```diff
@@ -0,0 +1,114 @@
+     * @param  {String} info.continuation   The continuation token, if any was passed
+     * @return {responseBody}
+     */
+    const handler = function handler(err, docs, info) {
```
Can we change this to filterHandler to be more explicit?

mkolt replied:
Can we change this to filterHandler to be more explicit? Same for the count sproc.

dwhieb replied:
https://github.com/Azure/azure-documentdb-js-server/pull/5/commits/d42afe467647fde8e3c40e12598b331f9889a95e

dwhieb replied:
https://github.com/Azure/azure-documentdb-js-server/pull/5/commits/d42afe467647fde8e3c40e12598b331f9889a95e

dwhieb commented:
@mkolt Should be all set!

mkolt commented on samples/stored-procedures/bulkImport.js:
```diff
@@ -1,59 +1,59 @@
-﻿/**
```
Is this file getting modified? Looks like it's not. Can we remove if it from the change?

mkolt commented on samples/stored-procedures/count.js:
```diff
@@ -1,66 +1,66 @@
-﻿/**
```
Is this file getting modified? Looks like it's not. Can we remove if it from the change?

mkolt commented on samples/stored-procedures/simpleScript.js:
```diff
@@ -1,26 +1,26 @@
-﻿/**
```
Does the change need to modify this file?

mkolt commented on samples/stored-procedures/count_underscoreAPI.js:
```diff
@@ -0,0 +1,87 @@
+    // filter the collection for documents using the filter object
+    const accepted = __.filter(function filter(doc) {
+
+      if (filterValue) {
```
Sorry :) Same perf issue. For optimized index, function calls and if-statements are currently not supported. How about if we change it to be like this: return filterKey === undefined || doc[filterKey] === doc[filterValue];
And same for the other sample...
